cmake/options.cmake

```diff
@@ -4,6 +4,54 @@
 # This source code is licensed in accordance with the terms specified in
 # the LICENSE file found in the root directory of this source tree.
 
+# Detect osquery version
+# The OSQUERY_VERSION cache variable will be used if set or not empty
+# Otherwise detect version through git and set it in the OSQUERY_VERSION_AUTODETECTED cache variable. If detection fails, 0.0.0 will be used.
+# Append OSQUERY_VERSION_SUFFIX cache variable to the final version if specified and the version is detected from git.
+# Verify if the final version is composed by three semver components, otherwise fail.
+# Returns the final version in OSQUERY_VERSION_INTERNAL and its components in OSQUERY_VERSION_COMPONENTS
+function(detectOsqueryVersion)
+  set(OSQUERY_VERSION "" CACHE STRING "Overrides osquery version with this value")
+  set(OSQUERY_VERSION_SUFFIX "" CACHE STRING "String to append when the version is automatically detected")
+  set(OSQUERY_VERSION_AUTODETECTED "" CACHE STRING "osquery version autodetected through git. Do not manually set." FORCE)
+  set(osquery_version 0.0.0)
+
+  if(NOT OSQUERY_VERSION)
+    find_package(Git REQUIRED)
+
+    execute_process(
+      COMMAND "${GIT_EXECUTABLE}" describe --tags --always --dirty
+      WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+      OUTPUT_VARIABLE branch_version
+      RESULT_VARIABLE exit_code
+    )
+
+    if(NOT ${exit_code} EQUAL 0)
+      message(WARNING "Failed to detect osquery version. Set it manually through OSQUERY_VERSION or 0.0.0 will be used")
+    else()
+      string(REGEX REPLACE "\n$" "" branch_version "${branch_version}")
+      set(osquery_version ${branch_version})
+      overwrite_cache_variable("OSQUERY_VERSION_AUTODETECTED" "STRING" ${osquery_version})
+
+      if(OSQUERY_VERSION_SUFFIX)
+        string(APPEND osquery_version "${OSQUERY_VERSION_SUFFIX}")
+      endif()
+    endif()
+  else()
+    set(osquery_version "${OSQUERY_VERSION}")
+  endif()
+
+  string(REPLACE "." ";" osquery_version_components "${osquery_version}")
+
+  list(LENGTH osquery_version_components osquery_version_components_len)
+  if(NOT osquery_version_components_len GREATER_EQUAL 3)
+    message(FATAL_ERROR "Version should have at least 3 components (semvar).")
+  endif()
+
+  set(OSQUERY_VERSION_INTERNAL "${osquery_version}" PARENT_SCOPE)
```

Comment:
This seems like a notable change. If I understand it correctly, any part of the build that references `OSQUERY_VERSION` would need to change to `OSQUERY_VERSION_INTERNAL`.

That seems fine to me, I imagine there might be a couple of iterations needed to get them all.

Reply:
That variable is only used by CMake and the only place where it's used has been already changed.

Reply:
Okay... I can't really parse how buck uses it, if at all. Or how it gets into the version table. But I'll take your word for it

Reply:
I've copied some snippets at the start of this PR that shows that: https://github.com/osquery/osquery/pull/5630#issuecomment-509753081

I'm not sure I fully understand what you're worried about but that's a CMake variable, it doesn't exist outside CMake scripts themselves, same for Buck.
It is used at a certain point to set three preprocessor defines (OSQUERY_VERSION, OSQUERY_BUILD_VERSION, OSQUERY_BUILD_SDK_VERSION), and those are then used in the code.